plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/MultiSigStateMachine.hs

```diff
@@ -46,19 +50,183 @@ import qualified Language.PlutusTx.StateMachine as SM
+-- | Equality of 'State' values
+stateEq :: State -> State -> Bool
+stateEq (InitialState v) (InitialState v') = Value.eq v v'
+stateEq (CollectingSignatures vl pmt pks) (CollectingSignatures vl' pmt' pks') =
+    P.and (P.and (Value.eq vl vl') (paymentEq pmt pmt')) (pkListEq pks pks')
+stateEq _ _ = False
+
+-- | @step params state input@ computes the next state given current state
+--   @state@ and the input.
+--   'step' does not perform any checks of the preconditions. This is done in
+--   'stepWithChecks' below.
+step :: State -> Input -> State
+step s i = case (s, i) of
+    (InitialState vl, ProposePayment pmt) ->
+        CollectingSignatures vl pmt []
+    (CollectingSignatures vl pmt pks, AddSignature pk) ->
+        CollectingSignatures vl pmt (pk:pks)
+    (CollectingSignatures vl _ _, Cancel) ->
+        InitialState vl
+    (CollectingSignatures vl (Payment vp _ _) _, Pay) ->
+        let vl' = Value.minus vl vp in
+        InitialState vl'
+    _ -> P.error (P.traceH "invalid transition" ())
+
+-- | @stepWithChecks params ptx state input@ computes the next state given
+--   current state @state@ and the input. It checks whether the pending
+--   transaction @ptx@ pays the expected amounts to script and public key
+--   addresses. Fails with 'P.error' if an invalid transition is attempted.
+stepWithChecks :: Params -> PendingTx -> State -> Input -> State
+stepWithChecks p ptx s i =
+    let newState = step s i in
+    case (s, i) of
+        (InitialState vl, ProposePayment pmt) ->
+            if isValidProposal vl pmt `P.and`
+                valuePreserved vl ptx
+            then newState
+            else P.traceErrorH "ProposePayment invalid"
+        (CollectingSignatures vl _ pks, AddSignature pk) ->
+            if Validation.txSignedBy ptx pk `P.and`
+                isSignatory pk p `P.and`
+                P.not (containsPk pk pks) `P.and`
+                valuePreserved vl ptx
+            then newState
+            else P.traceErrorH "AddSignature invalid"
+        (CollectingSignatures vl pmt _, Cancel) ->
+            if proposalExpired ptx pmt `P.and`
+                valuePreserved vl ptx
+            then InitialState vl
+            else P.traceErrorH "Cancel invalid"
+        (CollectingSignatures vl pmt@(Payment vp _ _) pks, Pay) ->
+            let vl' = Value.minus vl vp in
+            if P.not (proposalExpired ptx pmt) `P.and`
+                proposalAccepted p pks `P.and`
+                valuePreserved vl' ptx `P.and`
+                valuePaid pmt ptx
+            then newState
+            else P.traceErrorH "Pay invalid"
+        _ -> P.traceErrorH "invalid transition"
+
+mkValidator :: Params -> (State, Maybe Input) -> (State, Maybe Input) -> PendingTx -> ()
+mkValidator p ds vs ptx = 
+    let sm = StateMachine (stepWithChecks p ptx) stateEq in
+    SM.mkValidator sm ds vs ptx
+
+validator :: Params -> ValidatorScript
+validator params = 
+    ValidatorScript $ 
+        Ledger.applyScript
```

Comment:
I mildly think we should use `applyScript` infix throughout so the "function" comes first.

Reply:
Yeah, and it saves a few parens